utils/mcp_client.py

```
@@ -0,0 +1,531 @@
+    body = json.dumps(message, separators=(",", ":")).encode("utf-8")
+    return b"Content-Length: " + str(len(body)).encode("ascii") + b"\r\n\r\n" + body
+
+
+async def read_stdio_message(reader: asyncio.StreamReader) -> dict[str, Any]:
+    """Read one Content-Length framed JSON message from a subprocess stream."""
+    headers: dict[str, str] = {}
+
+    while True:
+        line = await reader.readline()
+        if line == b"":
+            raise McpClientError("MCP stdio server closed stdout")
```

Ruff (0.15.20)

[warning] Avoid specifying long messages outside the exception class

(TRY003)